airflow/models/taskinstance.py

```diff
@@ -819,17 +819,25 @@ def refresh_from_task(self, task: "Operator", pool_override=None):
         self.operator = task.task_type
 
     @provide_session
-    def clear_xcom_data(self, session=NEW_SESSION):
-        """
-        Clears all XCom data from the database for the task instance
+    def clear_xcom_data(self, session: Session = NEW_SESSION):
+        """Clear all XCom data from the database for the task instance.
+
+        If the task is unmapped, all XComs matching this task ID in the same DAG
+        run are removed. If the task is mapped, only the one with matching map
+        index is removed.
 
         :param session: SQLAlchemy ORM Session
         """
         self.log.debug("Clearing XCom data")
+        if self.map_index < 0:
+            map_index: Optional[int] = None
```

Comment:
why convert -1 to None?

Reply:
This is for clearing stale XComs; if the task instance is unmapped, there shouldn’t be mapped XComs bound to the task, so we use None to clear them all just in case.